### PR TITLE
[SYCL][InvokeSimd][E2E] Add new and fix existing InvokeSimd named barrier tests

### PR DESCRIPTION
**Pull request by fveselov**
Fixed:
* InvokeSimd/Regression/nbarrier_exec_in_order.cpp

Added new:
* InvokeSimd/Regression/nbarrier_loop.cpp
* InvokeSimd/Regression/nbarrier_multiple_wg.cpp

### CODE REVIEW

File: sycl/test-e2e/InvokeSimd/Regression/nbarrier_loop.cpp [hunk 1]
```diff
@@ -0,0 +1,212 @@
+// TODO: enable on Windows once driver is ready
+// NOTE: named barrier supported only since PVC
+// REQUIRES: gpu-intel-pvc && linux
+// UNSUPPORTED: cuda || hip
```
Comment (aelovikov-intel) [inline sycl/test-e2e/InvokeSimd/Regression/nbarrier_loop.cpp:4]: If https://github.com/intel/llvm/pull/9417 would get merged before this PR, this line won't be necessary.

File: sycl/test-e2e/InvokeSimd/Regression/nbarrier_exec_in_order.cpp [hunk 2]
```diff
@@ -127,27 +134,53 @@ bool test(QueueTY q) {
     out[i] = -1;
   }
 
+  auto dev = q.get_device();
+  auto deviceSLMSize =
+      dev.template get_info<sycl::info::device::local_mem_size>();
+  if constexpr (UseSLM)
+    std::cout << "Local Memory Size: " << deviceSLMSize << std::endl;
```
Comment (v-klochkov) [inline sycl/test-e2e/InvokeSimd/Regression/nbarrier_exec_in_order.cpp:141]: why only this printing is guarded by UseSLM check?

File: sycl/test-e2e/InvokeSimd/Regression/nbarrier_exec_in_order.cpp [hunk 3]
```diff
@@ -127,27 +134,53 @@ bool test(QueueTY q) {
...
   try {
     // workgroups
     sycl::range<1> GlobalRange{Size};
     // threads in each group
-    sycl::range<1> LocalRange{Threads};
+    sycl::range<1> LocalRange{Size};
     sycl::nd_range<1> Range{GlobalRange * LocalRange, LocalRange};
 
     auto e = q.submit([&](handler &cgh) {
+      auto LocalAcc = local_accessor<int, 1>(Size, cgh);
```
Comment (v-klochkov) [inline sycl/test-e2e/InvokeSimd/Regression/nbarrier_exec_in_order.cpp:159]: The creation and passing to ESIMDS of this local_accessor is done even with UseSLM=false.
Doesn't it look like a test design issue/problem to you?

File: sycl/test-e2e/InvokeSimd/Regression/nbarrier_exec_in_order.cpp [hunk 4]
```diff
@@ -41,7 +45,8 @@ constexpr int VL = 16;
 template <int case_num> class KernelID;
 
 template <unsigned Threads, unsigned Size, bool UseSLM>
-ESIMD_INLINE void ESIMD_CALLEE_nbarrier(int localID,
+ESIMD_INLINE void ESIMD_CALLEE_nbarrier(local_accessor<int, 1> LocalAcc,
```
Comment (v-klochkov) [inline sycl/test-e2e/InvokeSimd/Regression/nbarrier_exec_in_order.cpp:48]: This test is marked with "// XFAIL: gpu-intel-pvc"
Can you please add some details explaining what exactly causes the fail on this test? Is it caused by nbarrier handling or something else?

File: sycl/test-e2e/InvokeSimd/Regression/nbarrier_exec_in_order.cpp [hunk 5]
```diff
@@ -108,8 +106,17 @@ ESIMD_INLINE void ESIMD_CALLEE_nbarrier(int localID,
 
 template <unsigned Threads, unsigned Size, bool UseSLM>
 [[intel::device_indirectly_callable]] SYCL_EXTERNAL void __regcall SIMD_CALLEE_nbarrier(
+#ifdef USE_ACC_PTR
```
Comment (v-klochkov) [inline sycl/test-e2e/InvokeSimd/Regression/nbarrier_exec_in_order.cpp:109]: On the line 101 above is written:
```
  if constexpr (UseSLM) {
    auto res = experimental_esimd::lsc_slm_block_load<int, VL>(2 * off_slm);
    experimental_esimd::lsc_block_store<int, VL>(o + 2 * off, res);
  }
```
I maybe did not get the idea of the test, but that '2 *' part looks like an error. Perhaps, I am wrong. Shouldn't it be just 'off_slm' on L102 and 'o + off' on L103?

**Review by v-klochkov — COMMENTED** [review]
I reviewed only 1 of 3 files at this moment - the one that existed before.
@sarnex - can you please also review this PR including the 2 new files?

File: sycl/test-e2e/InvokeSimd/Regression/nbarrier_exec_in_order.cpp [hunk 6]
```diff
@@ -32,6 +32,10 @@
 #include <iostream>
 #include <type_traits>
 
+// TODO: When gpu driver can pass/accept accessor by value,
+// the work-around undef #ifdef US_ACC_PTR should be removed.
```
Comment (sarnex) [inline sycl/test-e2e/InvokeSimd/Regression/nbarrier_exec_in_order.cpp:36]:
```suggestion
// the work-around undef #ifdef USE_ACC_PTR should be removed.
```

File: sycl/test-e2e/InvokeSimd/Regression/nbarrier_exec_in_order.cpp (hunk 2 above)
Comment (sarnex) [reply sycl/test-e2e/InvokeSimd/Regression/nbarrier_exec_in_order.cpp:141]: +1, i would have expected the SLM error below to also be guarded.

File: sycl/test-e2e/InvokeSimd/Regression/nbarrier_loop.cpp [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+#include <sycl/ext/intel/experimental/esimd/memory.hpp>
+#include <sycl/ext/oneapi/experimental/invoke_simd.hpp>
+#include <sycl/sycl.hpp>
+
+#include <functional>
+#include <iostream>
+#include <type_traits>
+
+// TODO: When gpu driver can pass/accept accessor by value,
+// the work-around undef #ifdef US_ACC_PTR should be removed.
```
Comment (sarnex) [inline sycl/test-e2e/InvokeSimd/Regression/nbarrier_loop.cpp:36]:
```suggestion
// the work-around undef #ifdef USE_ACC_PTR should be removed.
```

File: sycl/test-e2e/InvokeSimd/Regression/nbarrier_loop.cpp [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+    bool is_producer = localID == indexes[j][0] || localID == indexes[j][1];
+    bool is_consumer = !is_producer;
+    // only-consumer or only-producer modes
+    unsigned int flag = is_producer ? 0x1 : 0x2;
+
+    unsigned int producers = prods[j];
+    unsigned int consumers = Threads - producers;
+
+    if (is_producer) {
+      unsigned int slm_store_off = j * sizeof(int) * slm_size / 4;
```
Comment (sarnex) [inline sycl/test-e2e/InvokeSimd/Regression/nbarrier_loop.cpp:85]: can you add some explanation to the constant division/multiplication? for ex i don't understand the /4 here.  if it's for a data type, can you use sizeof(datatype)?

File: sycl/test-e2e/InvokeSimd/Regression/nbarrier_loop.cpp [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+  } catch (sycl::exception const &e) {
+    std::cout << "SYCL exception caught: " << e.what() << '\n';
+    free(out, q);
+    return -1;
+  }
+
+  bool passed = true;
+  for (int i = 0; i < Size; i++) {
+    int etalon = 2;
+    if (i < Size / 4)
```
Comment (sarnex) [inline sycl/test-e2e/InvokeSimd/Regression/nbarrier_loop.cpp:191]: can we make the computation of etalon clearer? i don't understand this, i just see random calculations

File: sycl/test-e2e/InvokeSimd/Regression/nbarrier_multiple_wg.cpp [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+  bool is_producer = localID % 2 == 0;
+  bool is_consumer = !is_producer;
+  // only-producer or only-comsumer modes
+  unsigned int flag = is_producer ? 0x1 : 0x2;
+
+  // producer writes to SLM, consumer reads what producer wrote
+  unsigned int off_slm =
+      static_cast<uint32_t>(
+          reinterpret_cast<std::uintptr_t>(LocalAcc.get_pointer())) +
+      (is_producer ? global_off : (global_off - VL)) * sizeof(int);
```
Comment (sarnex) [inline sycl/test-e2e/InvokeSimd/Regression/nbarrier_multiple_wg.cpp:74]: is the difference between producer and consumer so we intentionally get overlapping regions? if so we can add a comment?

File: sycl/test-e2e/InvokeSimd/Regression/nbarrier_multiple_wg.cpp [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+  // signaling after data stored
+  __ESIMD_ENS::named_barrier_signal(bid, flag, producers, consumers);
+
+  if (is_consumer) {
+    // consumers waiting here for signal from producer
+    __ESIMD_ENS::named_barrier_wait(bid);
+    // read SLM and store to output
+    auto ret = experimental_esimd::lsc_slm_block_load<int, VL>(off_slm);
+    // store SLM to output
+    experimental_esimd::lsc_block_store<int, VL>(o + global_off - VL, ret);
```
Comment (sarnex) [inline sycl/test-e2e/InvokeSimd/Regression/nbarrier_multiple_wg.cpp:93]: can we make a var with a good name for `global_off - VL` and use it everywhere?

File: sycl/test-e2e/InvokeSimd/Regression/nbarrier_multiple_wg.cpp [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+    e.wait();
+  } catch (sycl::exception const &e) {
+    std::cout << "SYCL exception caught: " << e.what() << '\n';
+    free(out, q);
+    return false;
+  }
+
+  bool passed = true;
+  for (int i = 0; i < Size; i++) {
+    int etalon = (i / (2 * VL)) * 2;
```
Comment (sarnex) [inline sycl/test-e2e/InvokeSimd/Regression/nbarrier_multiple_wg.cpp:185]: also would help to have a comment explaining this

**Review by sarnex — COMMENTED** [review]
similar comments to Slava, some of his comments also apply to the new tests.

**Comment by aelovikov-intel** [comment]
Something is wrong with git history in this PR...

File: sycl/test-e2e/InvokeSimd/Regression/nbarrier_exec_in_order.cpp [hunk 13]
```diff
@@ -127,27 +134,53 @@ bool test(QueueTY q) {
...
-                sg.get_group_linear_id() * VL + g.get_linear_id() * Threads;
-            uint32_t wi_id = i + sg.get_local_id();
+
             // Thread local ID in ESIMD context
-            int localID = wi_id / VL;
+            int localID = sg.get_group_linear_id();
+
+            // SLM init
+            uint32_t slmID = item.get_local_id(0);
+            auto LocalAccCopy = LocalAcc;
```
[the remark this answers is not in the document]
Comment (fveselov) [reply sycl/test-e2e/InvokeSimd/Regression/nbarrier_exec_in_order.cpp]: I've made all the variable names more consistent

File: sycl/test-e2e/InvokeSimd/Regression/nbarrier_exec_in_order.cpp (hunk 2 above)
Comment (fveselov) [reply sycl/test-e2e/InvokeSimd/Regression/nbarrier_exec_in_order.cpp]: I decided to move this print outside of the _test_ function since the amount of available memory will never change between cases. I also added a guard to each case that uses SLM.

File: sycl/test-e2e/InvokeSimd/Regression/nbarrier_exec_in_order.cpp (hunk 4 above)
Comment (fveselov) [reply sycl/test-e2e/InvokeSimd/Regression/nbarrier_exec_in_order.cpp]: This test is fail with vISA LTO enabled and pass without the optimization.

File: sycl/test-e2e/InvokeSimd/Regression/nbarrier_exec_in_order.cpp (hunk 5 above)
Comment (fveselov) [reply sycl/test-e2e/InvokeSimd/Regression/nbarrier_exec_in_order.cpp:115]: This is the error, thanks!

File: sycl/test-e2e/InvokeSimd/Regression/nbarrier_exec_in_order.cpp (hunk 3 above)
Comment (fveselov) [reply sycl/test-e2e/InvokeSimd/Regression/nbarrier_exec_in_order.cpp]: As a workaround, I created a dummy local_accessor with a size of 0 and added guards wherever SLM was used. Is this acceptable?

File: sycl/test-e2e/InvokeSimd/Regression/nbarrier_loop.cpp (hunk 8 above)
Comment (fveselov) [reply sycl/test-e2e/InvokeSimd/Regression/nbarrier_loop.cpp]: I reworked this part a bit and added some comments. Hope now it is more clear what's going on in the test.

File: sycl/test-e2e/InvokeSimd/Regression/nbarrier_loop.cpp (hunk 9 above)
Comment (fveselov) [reply sycl/test-e2e/InvokeSimd/Regression/nbarrier_loop.cpp]: Also reworked and with some comments

File: sycl/test-e2e/InvokeSimd/Regression/nbarrier_multiple_wg.cpp (hunk 12 above)
Comment (fveselov) [reply sycl/test-e2e/InvokeSimd/Regression/nbarrier_multiple_wg.cpp]: I have added some comments to all the places mentioned above and other places throughout the tests. Let me know how I can further improve the readability of the tests.

File: sycl/test-e2e/InvokeSimd/Regression/nbarrier_exec_in_order.cpp [hunk 14]
```diff
@@ -157,14 +208,14 @@ bool test(QueueTY q) {
 
   bool passed = true;
   for (int i = 0; i < Size; i++) {
-    int etalon = i * 2 * Threads / Size;
-    if (etalon == Threads) // last stored chunk
-      etalon -= 1;
-    if (etalon > Threads) // excessive part of surface
-      etalon = 0;
-    if (out[i] != etalon) {
+    int ref = 2 * i * Threads / Size;
```
Comment (sarnex) [inline sycl/test-e2e/InvokeSimd/Regression/nbarrier_exec_in_order.cpp:211]: this is the global buffer write you were talking about right? are we just doubling the offset to guarantee no overlap?

File: sycl/test-e2e/InvokeSimd/Regression/nbarrier_multiple_wg.cpp [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+  } catch (sycl::exception const &e) {
+    std::cout << "SYCL exception caught: " << e.what() << '\n';
+    free(out, q);
+    return false;
+  }
+
+  bool passed = true;
+  for (int i = 0; i < Size; i++) {
+    int ref = i / VL;
+    if (ref % 2 == 1) // every odd ref is skipped
```
Comment (sarnex) [inline sycl/test-e2e/InvokeSimd/Regression/nbarrier_multiple_wg.cpp:182]: is that because the off refs are the producers doing an slm write and not a global buffer write?

File: sycl/test-e2e/InvokeSimd/Regression/nbarrier_basic.cpp [hunk 16]
```diff
@@ -56,7 +56,7 @@ ESIMD_INLINE void ESIMD_CALLEE_nbarrier(nd_item<1> *ndi) SYCL_ESIMD_FUNCTION {
     nd_item<1> *ndi) SYCL_ESIMD_FUNCTION;
 
 int main(void) {
-  auto Queue = queue{gpu_selector_v};
```
Comment (sarnex) [inline sycl/test-e2e/InvokeSimd/Regression/nbarrier_basic.cpp:59]: do we really not need the selector anymore?

**Review by sarnex — COMMENTED** [review]
a lot easier to understand, thanks!

File: sycl/test-e2e/InvokeSimd/Regression/nbarrier_basic.cpp (hunk 16 above)
Comment (fveselov) [reply sycl/test-e2e/InvokeSimd/Regression/nbarrier_basic.cpp:59]: Test works exactly the same with or without it, I tried both ways.

File: sycl/test-e2e/InvokeSimd/Regression/nbarrier_exec_in_order.cpp [hunk 17]
```diff
@@ -49,76 +54,99 @@ ESIMD_INLINE void ESIMD_CALLEE_nbarrier(int localID,
...
+   * and so on, but last thread skips this block
+   */
+  if (local_id < Threads - 1) {
+    int barrier_id = local_id + 1;
     __ESIMD_ENS::named_barrier_signal(barrier_id, flag, producers, consumers);
     __ESIMD_ENS::named_barrier_wait(barrier_id);
   }
 
+  /* Here we wait for all threads to sync, and each thread now copies the
+   * non-overlapping region from the SLM to the global buffer, so the offset is
```
[the remark this answers is not in the document]
Comment (fveselov) [reply sycl/test-e2e/InvokeSimd/Regression/nbarrier_exec_in_order.cpp]: This particular part is just to copy the content of SLM to the global buffer for further self-check purpose. Overlapping store is above, I added more comments to point out the actual payload instruction.

File: sycl/test-e2e/InvokeSimd/Regression/nbarrier_exec_in_order.cpp [hunk 18]
```diff
@@ -127,25 +155,48 @@ bool test(QueueTY q) {
 
   try {
     // workgroups
-    sycl::range<1> GlobalRange{Size};
```
[the remark this answers is not in the document]
Comment (fveselov) [reply sycl/test-e2e/InvokeSimd/Regression/nbarrier_exec_in_order.cpp:130]: It is only to make all variable names consistent throughout the test.

File: sycl/test-e2e/InvokeSimd/Regression/nbarrier_exec_in_order.cpp (hunk 14 above)
Comment (fveselov) [reply sycl/test-e2e/InvokeSimd/Regression/nbarrier_exec_in_order.cpp]: I added a comment and changed the calculation slightly, but the resulting value remained the same.
TLDR: Reference value is the ID of the thread that stored it. To calculate it, we need to divide **_i_** by **_(VL / 2)_** and it comes from the fact that threads are writing their values with **_(VL / 2)_** offsets from each other.

File: sycl/test-e2e/InvokeSimd/Regression/nbarrier_multiple_wg.cpp (hunk 15 above)
Comment (fveselov) [reply sycl/test-e2e/InvokeSimd/Regression/nbarrier_multiple_wg.cpp]: Added a comment to this part as well.
Since the producers have only even thread IDs and are the only ones doing a write, any odd ref must be skipped.

File: sycl/test-e2e/InvokeSimd/Regression/nbarrier_exec_in_order.cpp (hunk 3 above)
Comment (v-klochkov) [reply sycl/test-e2e/InvokeSimd/Regression/nbarrier_exec_in_order.cpp]: Thank you. It is better, but still odd because UseSLM=false cases would have to be disabled if something happens with passing local accessors (e.g. regression in gpu rt).
I am not insisting on splitting the tests though as it is code duplication. It is up to you.

**Comment by fveselov** [comment]
SYCL Pre Commit / Linux / OCL x64 LLVM Test Suite (pull_request_target) fail unrelated to changes:
```
Failed Tests (1):
  SYCL :: Scheduler/DataMovement.cpp
```

**Review by v-klochkov — APPROVED** [review]
